### ISSUE TITLE
[Proposal] Add Subtitle Support to Media Element

### ISSUE THREAD

**Issue by ne0rrmatrix** [CommunityToolkit/Maui#1899]
### Feature name

Media Element Subtitles

### Link to discussion

https://github.com/CommunityToolkit/Maui/discussions/1867

### Progress tracker

- [X] Android Implementation
- [X] iOS Implementation
- [X] MacCatalyst Implementation
- [X] Windows Implementation
- [ ] Tizen Implementation
- [ ] Unit Tests
- [ ] Samples
- [ ] Documentation

### Summary

Add support for Subtitles to Media Element. Initial support for Windows, Mac, iOS, and Android. Subtitle format allowed at launch include srt and sub.

### Windows ###
![windows-subtitle-normal](https://github.com/CommunityToolkit/Maui/assets/4167863/7b5989cb-27f2-4bb7-8a2e-b2b219d151ff)

![Windows-subtitle-fullscreen](https://github.com/CommunityToolkit/Maui/assets/4167863/8e4f022a-4948-41cd-8aa8-e64c7a1f7ca7)

### Android ###

![Android-normal-subtitles](https://github.com/CommunityToolkit/Maui/assets/4167863/c331279c-f230-4bbb-a6b3-78149d0b621a)

![Android-subtitles-fullscreen](https://github.com/CommunityToolkit/Maui/assets/4167863/82e1deda-11aa-4640-b200-9e510f12a692)

### IOS ###
https://github.com/CommunityToolkit/Maui/assets/4167863/ab3c3f09-ddb4-4ba4-a8db-f84efba15ab9

### Motivation

I would like to help support people who have difficulty with hearing or need assistance in loud noisy environments where it is hard to hear the dialogue. Or people that want translations of foreign languages spoken that they do not know. Other people just like subtitles. This will add robust support that can be expanded later.

### API: ###
```
	/// <summary>
	/// Backing store for the <see cref="SubtitleFont"/> property.
	/// </summary>
	public static readonly BindableProperty SubtitleFontProperty = BindableProperty.Create(nameof(SubtitleFont), typeof(string), typeof(MediaElement), string.Empty);
	
	/// <summary>
	/// Backing store for the <see cref="SubtitleFontSize"/> property.
	/// </summary>
	public static readonly BindableProperty SubtitleFontSizeProperty = BindableProperty.Create(nameof(SubtitleFontSize), typeof(double), typeof(MediaElement), 16.0);

	/// <summary>
	/// Backing store for the <see cref="SubtitleUrl"/> property.
	/// </summary>
	public static readonly BindableProperty SubtitleProperty = BindableProperty.Create(nameof(SubtitleUrl), typeof(string), typeof(MediaElement), string.Empty);
```

### Usage Syntax

```markdown
### XAML Usage:


 <toolkit:MediaElement
     x:Name="MediaElement"
     ShouldAutoPlay="True"
     SubtitleFont="monospace"
     SubtitleFontSize="20"
     SubtitleUrl=""
     Source="https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4"/>


### C# Code behind example ###
```csharp
MediaElement.SubtitleFont = "monospace";
MediaElement.SubtitleFontSize = 16;
MediaElement.SubtitleUrl = "https://raw.githubusercontent.com/ne0rrmatrix/SampleVideo/main/SRT/WindowsVideo.srt";
MediaElement.Source = MediaSource.FromResource("WindowsVideo.mp4");
```
```


### Drawbacks

Currently it uses a custom class to read the subtitle files and convert them to strings that can be used at specific times in a text box appropriate for each OS. This is a new class that only supports plain text at the moment. Non standard files and RTF formatted files are not supported. 

### Alternatives

The custom class for reading subtitle files was written to allow iOS and Mac Catalyst support as the current methods are to use m3u8 files and do not support directly linking a subtitle file. ATM we have support for m3u8 files but we have no control over how or what is displayed at all. Android and Windows both have robust libraries that support subtitles. They do not support any common format of subtitles.

ATM if you want uniform support for multiple different formats on each and every device my custom class is the way to go.

### Unresolved Questions

How are we going to use/add fonts? Are we going to go the normal route and just add fonts to the project and import them like any other maui project? Are we going to just use built in supported OS fonts for each device?

**Comment by ne0rrmatrix** [comment]
An example where I could use help is getting the subtitles in the in right position vertically in android. I have been unable to get that working at all when switching to full screen. If anyone is an android expert with native UI can you look at the code and make a suggestion for how to get set the vertical offset from the bottom and not have it stuck at the top. It works fine when not in full screen. But once you switch it does not work and is stuck at the top.

**Comment by ne0rrmatrix** [comment]
The issue with android has been resolved. New issue I need help with is how to use fonts in a nuget package loaded in sample app. ATM this is holding back further development. I see no way to access a font loaded in an app by a nuget package loaded in that app. If anyone has any suggestions I would appreciate it. Otherwise maybe adding it after we finish v1 as an addition to API at a later date?

**Comment by ne0rrmatrix** [comment]
All the work for this features current API is done. I have been refactoring code. I need to update the docs and get them approved.  In about another week I will mark it as ready for review. I need to do device testing to see if there are any show stopper bugs or things that emulators missed.

**Comment by imememani** [comment]
How's this looking? Feel like this could be a really great addition.

**Comment by JonathanWebsdale** [comment]
Any updates on availability? Currently investigating the use of bunny.net as a media provider and restricted to using their viewer hosted in an iFrame inside a WebView control (which brings its own problems), but they do have support for VVT and SRT format subtitle files.  Would ideally want to test this within the MediaElement.

**Comment by bijington** [comment]
I believe @ne0rrmatrix was really close with this. We have piled a lot of effort into .NET 9.0 support which we hope to merge in the next week or so and then we can start to revisit open PRs like the one for this proposal 👍

**Comment by ne0rrmatrix** [comment]
The last unresolved issue was solved months ago. I will update that this morning to reflect that I fixed the font issue. The one thing I have done that will require input is a complete rewrite that implements all this by using built in classes and we need to decide whether we go with custom implementation for subtitles or use native methods. Using native libs removes the need to support it but with custom libs we can do anything we want like allow third party support for custom subtitle support. An example is you can insert your own implementation for parser with this version but native libs don't allow this.

**Comment by MuhammadTalhaahmed** [comment]
> The last unresolved issue was solved months ago. I will update that this morning to reflect that I fixed the font issue. The one thing I have done that will require input is a complete rewrite that implements all this by using built in classes and we need to decide whether we go with custom implementation for subtitles or use native methods. Using native libs removes the need to support it but with custom libs we can do anything we want like allow third party support for custom subtitle support. An example is you can insert your own implementation for parser with this version but native libs don't allow this.

You mean Subtitle support is finally ready and is going to merge in main ?
Can you explain it more simply what you said about "Custom implementation for Subtitles" or "Native Methods".

**Comment by ne0rrmatrix** [comment]
You will be able to add a custom parser and support any text based subtitles that media element does not itself natively support. ATM I have added support for srt and vtt files. Or if you find a need for say supporting html and do not want to use my text only srt or vtt you will be able to just your own implementation and integrate that with media element directly yourself. It has great versitility. The choice will be yours. Or we can go with native libs(I have both versions ready) which will have better UI support with native buttons on windows and android plus support mulitple languages in both but not in IOS or mac os. Both versions do support both vtt and srt in a single file. 

The key differences are the native version supports a UI button and both versions let you select a variety of subtitle files that can be added as a simple list of URL's for both windows and android. This is not supported on apple devices natively. On the other hand you can write your own parser to fix any issues mine has plus you can add additional supported file types as you want with my custom implementation.  I will let the community decide which version they want.
